imap_processing/mag/l1c/mag_l1c.py

```diff
@@ -39,19 +51,444 @@ def mag_l1c(
+    1. find all the gaps in the dataset
+    2. generate a new timeline with the gaps filled
+    3. fill the timeline with normal mode data (so, all the non-gap timestamps)
+    4. interpolate the gaps using the burst mode data and the method specified in
+        interpolation_function.
+
+    It returns an (n, 8) shaped array:
+    0 - epoch (timestamp)
+    1-4 - vector x, y, z, and range
+    5 - generated flag (0 for normal data, 1 for interpolated data, -1 for missing data)
+    6-7 - compression flags (is_compressed, compression_width)
+
+    Parameters
+    ----------
+    normal_mode_dataset : xarray.Dataset
+        The normal mode dataset, which acts as a base for the output.
+    burst_mode_dataset : xarray.Dataset
+        The burst mode dataset, which is used to fill in the gaps in the normal mode.
+    interpolation_function : InterpolationFunction
+        The interpolation function to use to fill in the gaps.
+
+    Returns
+    -------
+    np.ndarray
+        An (n, 8) shaped array containing the completed timeline.
+    """
+    norm_epoch = normal_mode_dataset["epoch"].data
+    vecsec_attr = normal_mode_dataset.attrs["vectors_per_second"]
+
+    output_dataset = normal_mode_dataset.copy(deep=True)
+    output_dataset["sample_interpolated"] = xr.DataArray(
+        np.zeros(len(normal_mode_dataset))
+    )
+
+    gaps = find_all_gaps(norm_epoch, vecsec_attr)
+
+    new_timeline = generate_timeline(norm_epoch, gaps)
+    norm_filled = fill_normal_data(normal_mode_dataset, new_timeline)
+    interpolated = interpolate_gaps(
+        burst_mode_dataset, gaps, norm_filled, interpolation_function
+    )
+
+    return interpolated
+
+
+def fill_normal_data(
+    normal_dataset: xr.Dataset, new_timeline: np.ndarray
+) -> np.ndarray:
+    """
+    Fill the new timeline with the normal mode data.
+
+    If the timestamp exists in the normal mode data, it will be filled in the output.
+
+    Parameters
+    ----------
+    normal_dataset : xr.Dataset
+        The normal mode dataset.
+    new_timeline : np.ndarray
+        A 1D array of timestamps to fill.
+
+    Returns
+    -------
+    np.ndarray
+        An (n, 8) shaped array containing the timeline filled with normal mode data.
+        Gaps are marked as -1 in the generated flag column at index 5.
+        Indices: 0 - epoch, 1-4 - vector x, y, z, and range, 5 - generated flag,
+        6-7 - compression flags.
+    """
+    # TODO: fill with FILLVAL?
+    filled_timeline: np.ndarray = np.zeros((len(new_timeline), 8))
+    filled_timeline[:, 0] = new_timeline
+    # Flags, will also indicate any missed timestamps
+    filled_timeline[:, 5] = -1
+
+    for index, timestamp in enumerate(normal_dataset["epoch"].data):
+        timeline_index = np.searchsorted(new_timeline, timestamp)
+        filled_timeline[timeline_index, 1:5] = normal_dataset["vectors"].data[index]
+        filled_timeline[timeline_index, 5] = 0
```

Comment:
```suggestion
        filled_timeline[timeline_index, 5] = L1C_DATA_ORIGIN.FROM_NORMAL_MODE
```
Be tempted to put some of the magic strings and integers into constants/enums to make the intention a bit clearer to the reader